src/net_processing.cpp

```diff
@@ -1642,6 +1659,20 @@ void PeerManagerImpl::FinalizeNode(const CNode& node)
     LogPrint(BCLog::NET, "Cleared nodestate for peer=%d\n", nodeid);
 }
 
+bool PeerManagerImpl::HasAllDesirableServiceFlags(ServiceFlags services) const
+{
+    // Shortcut for (services & GetDesirableServiceFlags(services)) == GetDesirableServiceFlags(services)
+    return !(GetDesirableServiceFlags(services) & (~services));
+}
+
+ServiceFlags PeerManagerImpl::GetDesirableServiceFlags(ServiceFlags services) const
+{
+    if (services & NODE_NETWORK_LIMITED) {
+        if (m_initial_sync_finished) return ServiceFlags(NODE_NETWORK_LIMITED | NODE_WITNESS);
+    }
+    return ServiceFlags(NODE_NETWORK | NODE_WITNESS);
```

Comment:
Can we return `StatelessServiceFlags()` here instead?

Reply:
Yeah this what causes my confusion [here](https://github.com/bitcoin/bitcoin/pull/28170#discussion_r1376837397) as well. I think the author means to use `StatelessServiceFlags` only for seed peers. In that case, i'd probably say that explicitly in the function's name. Like `SeedsServiceFlags`. Then, a comment saying something like `these flags should probably be in sync with GetDesirableServiceFlags, but not strictly`.

Reply:
> In that case, i'd probably say that explicitly in the function's name. Like SeedsServiceFlags.

Sounds good. Will rename it to `SeedsServiceFlags`.
I was thinking we would use this function for other cases in the future, but agree that this can be simplified for now.

> Then, a comment saying something like these flags should probably be in sync with GetDesirableServiceFlags, but not strictly.

I'm not sure about this comment. We could have a much more complex logic at the PeerManager level, adaptable based on the chain context and/or the user's configurations. For instance, the user could be running a node whose preferred connections are ones supporting `NODE_P2P_V2` or `NODE_COMPACT_FILTERS` (because the sync mechanism is BIP157) or some other service flag in the future.